### DISCUSSION TITLE
difference between glsp and sprotty

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-glsp/glsp#1104] [Q&A]
recently,i learn about eclipse glsp and sprotty,the more  i leraned ,the more i coufused, glsp based on sprotty,i want to know since there has sprotty,what is the meaning of glsp framework？

**Comment by rhumbertgz** [comment]
GLSP extends Sprotty based diagram with editing functionality, extra component abstractions, and a communication mechanism with the server based on the LSP.

**Reply by zrwsmd** [reply]
thanks for your apply

**Comment by planger** [comment]
Thank you for your interest in Eclipse GLSP and Eclipse Sprotty!

As you say, GLSP uses Sprotty on the client for rendering diagrams. On top of Sprotty, GLSP adds several aspects
- On the client: additional capabilities for diagram editing and other additional features, such as copy/paste, undo/redo, diagram navigation, etc.
- An extensible client-server protocol for communicating with a GLSP server, such as requesting changes to be applied to the underlying model(s), triggering validation of the model, etc.
- Generic server frameworks (for Java and Typescript), which implement the diagram logic and the management of the underlying model.
- Tool platform integrations, to integrate GLSP editors into VS Code, Eclipse Theia, Eclipse RCP, or plain web applications.

While there are overlapping use cases that could be implemented with either GLSP and Sprotty, the focus of those two projects is orthogonal / different:
GLSP focuses on enabling the implementation of fully editable "model-first" diagram editors: the diagram editor is the primary editor for users to interact with a model.
Sprotty on its own focuses on providing the basis for implementing auxiliary, interactive views visualizing other data sources. A typical use case in that area is to provide an auxiliary view for "text-first" models, e.g. to visualize aspects, such as relationships among elements of a textual model. In those use cases, the user mainly edits a textual file and on change of this file, a secondary views based on Sprotty is updated to visualize the underlying model of the textual file (see [a blog post](https://www.typefox.io/blog/langium-meets-sprotty-combining-text-and-diagrams-in-vs-code) on this use case). As a consequence, Sprotty provides only the UI basis for diagram editing, but doesn't put this aspect into its center as often it is used in combination with another textual editor, which serves as the main editor for users. GLSP on the other hand provides a bunch of generic editing tools (such as adding, deleting, changing, resizing elements, etc), which communicate those changes to the server, which in turn applies those changes to the underlying model, as fully editable "model-first" diagram editors are in its focus.

So if you mainly are looking for a framework to create an interactive view visualizing a local or remote data source graphically, Sprotty is usually sufficient and thus the right choice. If you want to combine such an interactive view with a textual model, combining Sprotty with Langium or Xtext is a very nice combination.
If you want to build a full fledged diagram editor, whereas the diagram is the user's main artifact, GLSP might be the better choice, as you benefit from GLSP's editing tools and server frameworks for managing the underlying model.

I hope this helps!

**Reply by zrwsmd** [reply]
thanks for your knowledgable apply ,i learn a lot from your article

**Reply by zrwsmd** [reply]
if i want to achievement a requirement when diargram changed(right part),corresponding left part also synchronized change too,if i use custom source model like below picture with extension .sm rather than EMF model or json or xml,is it possible to achieve by glsp?sprotty achieved this requirement,but i can not find related code about how it achieved
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/49bfeaae-99fc-49cc-84eb-e47439d6d9ae)

**Reply by planger** [reply]
The left part is independent of GLSP and needs to be implemented separately and integrated in a custom integration with GLSP. Usually you'd use a framework such as [Langium](https://langium.org/) or [Xtext](https://eclipse.dev/Xtext/) for such a textual language and implement the source model of GLSP to interact with the abstract syntax tree parsed from the left text file (via the parser created from Langium or Xtext). On change in the GLSP diagram editor, you'd update the abstract syntax tree and serialize it to text and update the left editor.
Note that there are some challenges in doing that (e.g. loss of whitespace after a change in the diagram, need for autolayouting in the diagram after a change in the textual file, two different editing levels [textual character changes in text file vs model editing commands in diagram], etc.).
This is why most of the implementations that provide a graphical and a textual editor either try to set them up so that the graphical view is mostly readonly and generated (see e.g. [Langium+Sprotty](https://www.typefox.io/blog/langium-meets-sprotty-combining-text-and-diagrams-in-vs-code)) or that both editors edit a disjunct part of the underyling model.

**Reply by zrwsmd** [reply]
for Langium+Sprotty example,how it achieve when diargram changed(right part),corresponding left part also synchronized change too,any code refer to?i just did not find it,It's too abstract to understand

**Reply by planger** [reply]
In the Langium-Sprotty example, the graphical part is readonly, so it only changes if you change the textual representation on the left. You could add editing there and translate those into corresponding textual changes.
I'm not aware of more sophisticated / complete public examples. The ones I know are closed-source products, I'm afraid.

**Reply by zrwsmd** [reply]
how to change abstract syntax tree when on change in the GLSP diagram editor,can you use langium for example?

**Reply by planger** [reply]
Here is a project that uses this approach: https://github.com/CrossBreezeNL/crossmodel

It is a little bit involved to set everything up so that this works and hard to explain in a single comment. But the idea is that langium parses the text into an abstract syntax tree and GLSP uses that as its source model, and on change in the diagram editor, it changes the abstract syntax tree model, serializes it and updates the text state in the langium language server.

We have a [talk at EclipseCon](https://www.eclipsecon.org/2023/sessions/building-cloud-native-modeling-tools) on this topic and will publish an example tool in October or so to demonstrate this approach.

**Reply by zrwsmd** [reply]
is sprotty-langium similar to glsp-langium?to change abstract syntax tree when a change in the diagram editor?sprotty seems no source model or graphical model concepts,if not ,how it achieved,i just found https://github.com/eclipse-sprotty/sprotty-vscode/blob/master/packages/sprotty-vscode-webview/src/lsp/editing/delete-with-workspace-edit.ts,the code seems to have nothing to do with operationg ast tree

**Reply by planger** [reply]
No, I think sprotty-langium doesn't focus on supporting changes done via the graphical representation and if, they rather suggest to trigger textual code changes in the textual view, which makes a lot of sense and manages the complexity. After all it is a trade-off between complexity and whether or not sophisticated changes in the graphical view are to be supported.

**Reply by zrwsmd** [reply]
but old sprotty-xtext(https://github.com/eclipse-sprotty/sprotty-vscode/issues/93) truly can achieve this requirement(when diagram change,corresponding text file  changes),code is here https://github.com/eclipse-sprotty/sprotty-vscode/blob/master/packages/sprotty-vscode-webview/src/lsp/editing/delete-with-workspace-edit.ts,i just did not catch how it works ,code is a little confused

**Reply by planger** [reply]
Yes, you can achieve that from both directions (sprotty-xtext/-langium having a text-first approach or glsp-xtext/-langium having rather a model-oriented approach). Depending on the requirements, both approaches come with their own challenges. Additionally you'll have the additional challenges of combining textual/graphical with both approaches too (e.g. whitespace handling vs layouting, different editing paradigm between textual and graphical, etc.).

**Comment by zrwsmd** [comment]
my requiremnet is when user performs a change in the diagram(my requirement diagram  is industrial Programming Language LD(Ladder Diagram)) ,corresponding lauguage like C++ or PLC ST (Structured Text）language generated,which combination is more suitable,
sprotty-xtext/-langium or glsp-xtext/-langium,

**Reply by planger** [reply]
The main question is whether changing the ST code should update the ladder diagram. If this is not much in the focus, I'd go for GLSP and just plainly generate the ST code after each change from the GLSP editor. Otherwise, probably both is fine, whereas if there are complex edit operations in the diagram, I'd still go for GLSP.
If the main editor for the user is the textual ST, I'd go for Sprotty.

**Reply by zrwsmd** [reply]
my requirement is when update the ladder diagram , corresponding change the ST code .
i am now thinking use glsp-langium as you said before ,use ast tree as glsp source model,when change in the diagram editor, it changes the abstract syntax tree model.
i want to achieve by two steps,first step is when update the ladder diagram,corresponding ast tree also changed,second step is ST 
code can generate from ast tree,i can refer to https://github.com/CrossBreezeNL/crossmodel as my requirement example,am i right?

**Reply by planger** [reply]
Yes, in crossmodel the AST is updated and code is being generated after each change.

**Reply by zrwsmd** [reply]
thanks a lot

**Comment by zrwsmd** [comment]
i notice https://github.com/CrossBreezeNL/crossmode project seem to have many source model file  which in separate different .cm files,
is SystemDiagram mainly used for render on the diagram  and i can regard it as an abstract semantic describe for Entity and Relationship,
am i right?

**Reply by planger** [reply]
That's right!

**Reply by zrwsmd** [reply]
thanks a lot
